pkg/dependencies/helmfactory.go

```diff
@@ -0,0 +1,138 @@
+)
+
+// ExecutableBuilder builds the Helm executeble and returns it.
+type ExecutableBuilder interface {
+	BuildHelmExecutable(...executables.HelmOpt) *executables.Helm
+}
+
+// HelmFactoryOpt configures a HelmFactory.
+type HelmFactoryOpt func(*HelmFactory)
+
+// HelmFactory is responsible for creating and owning instances of Helm client.
+type HelmFactory struct {
+	client         cluster.Client
+	helmClient     *cluster.HelmClient
+	mu             sync.Mutex
+	builder        ExecutableBuilder
+	registryMirror *registrymirror.RegistryMirror
+	env            map[string]string
+	insecure       bool
+}
+
+// WithRegistryMirror configures the factory to use registry mirror wherever applicable.
+func WithRegistryMirror(registryMirror *registrymirror.RegistryMirror) HelmFactoryOpt {
+	return func(hf *HelmFactory) {
+		hf.registryMirror = registryMirror
+	}
+}
+
+// WithEnv configures the factory to use proxy configurations wherever applicable.
+func WithEnv(env map[string]string) HelmFactoryOpt {
+	return func(hf *HelmFactory) {
+		hf.env = env
+	}
+}
+
+// WithInsecure configures the factory to configure helm to use to allow connections to TLS registry without certs or with self-signed certs.
+func WithInsecure() HelmFactoryOpt {
+	return func(hf *HelmFactory) {
+		hf.insecure = true
+	}
+}
+
+// NewHelmFactory returns a new HelmFactory.
+func NewHelmFactory(client cluster.Client, builder ExecutableBuilder, opts ...HelmFactoryOpt) *HelmFactory {
+	hf := &HelmFactory{
+		client:  client,
+		builder: builder,
+		mu:      sync.Mutex{},
+	}
+
+	for _, o := range opts {
+		o(hf)
+	}
+
+	return hf
+}
+
+// GetClient returns a new Helm executeble.
+func (f *HelmFactory) GetClient(opts ...executables.HelmOpt) cluster.Helm {
+	defaultOpts := []executables.HelmOpt{}
+
+	if f.registryMirror != nil {
+		defaultOpts = append(defaultOpts, executables.WithRegistryMirror(f.registryMirror))
+	}
+	if f.env != nil {
+		defaultOpts = append(defaultOpts, executables.WithEnv(f.env))
+	}
+
+	if f.insecure {
+		defaultOpts = append(defaultOpts, executables.WithInsecure())
+	}
+
+	opts = append(defaultOpts, opts...)
+	return f.builder.BuildHelmExecutable(opts...)
+}
+
+// GetClientForCluster returns a new Helm client configured using information from the provided cluster's management cluster.
+func (f *HelmFactory) GetClientForCluster(ctx context.Context, clus *anywherev1.Cluster) (*cluster.HelmClient, error) {
+	managmentCluster := clus
+
+	var rUsername, rPassword string
+	var err error
+
+	if clus.IsManaged() {
+		managmentCluster = &anywherev1.Cluster{}
+		if err := f.client.Get(ctx, clus.ManagedBy(), clus.Namespace, managmentCluster); err != nil {
+			return nil, err
+		}
+	}
+
+	if managmentCluster.RegistryAuth() {
+		rUsername, rPassword, err = f.getClusterRegistryCredentrials(ctx, managmentCluster)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	r := registrymirror.FromCluster(managmentCluster)
+	helm := f.GetClient(executables.WithRegistryMirror(r))
```

Comment:
does this work if there is no registry mirror config?

Reply:
Yes, `registrymirror.FromCluster` would return nil and then the option would simply set it to nil on the `Helm` instance.

Reply:
ok
maybe it's worth doing an if
since this assumes that creating a client with a nil registr5y mirror is valid
but kind of nit